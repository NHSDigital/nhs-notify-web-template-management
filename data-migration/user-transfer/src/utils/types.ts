export type Parameters = {
  environment: string;
  component: string;
  accessKeyId: string;
  secretAccessKey: string;
  userPoolId: string;
  region: string;
  sessionToken: string;
  flag?: string;
};

export type UserData = {
  username: string;
  /*
   * extracted from the user's group starting with 'client:'
   */
  clientId: string;
  /*
   * extracted from the user's 'sub' attribute
   */
  userId: string;
};

export type Template = {
  id: string;
  owner: string;
};

type ToFrom = {
  from: string;
  to: string;
};

export type TransferStage =
  | 's3:copy'
  | 'ddb:transfer'
  | 's3:delete'
  | 'initial'
  | 'finished';

export type TransferStatus = 'success' | 'failed' | 'migrate';

export type UserTransferPlanItem = {
  templateId: string;
  status: TransferStatus;
  stage: TransferStage;
  reason?: string;
  ddb: {
    owner: ToFrom;
  };
  s3: {
    files: ToFrom[];
  };
};

export type UserTransferPlan = {
  total: number;
  tableName: string;
  bucketName: string;
  migrate: {
    count: number;
    plans: UserTransferPlanItem[];
  };
  orphaned: {
    count: number;
    templateIds: string[];
  };
};

export type UserTransferPlanItemResult = {
  success: boolean;
  stage: TransferStage;
  reasons?: string[];
};
