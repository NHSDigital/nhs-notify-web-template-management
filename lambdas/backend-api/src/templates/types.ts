export type User = {
  userId: string;
  clientId: string | undefined;
};
