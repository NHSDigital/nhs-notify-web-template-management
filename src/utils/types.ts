export type FormId =
  | 'choose-a-template-type'
  | 'create-nhs-app-template'
  | 'create-email-template'
  | 'create-letter-template'
  | 'create-text-message-template'
  | 'preview-nhs-app-template'
  | 'submit-template'
  | 'create-nhs-app-template-back';

export type FormErrorState = {
  formErrors: string[];
  fieldErrors: Record<string, string[]>;
};

export enum TemplateType {
  NHS_APP = 'NHS_APP',
  SMS = 'SMS',
  EMAIL = 'EMAIL',
  LETTER = 'LETTER',
}

export type Session = {
  id: string;
  templateType: TemplateType | 'UNKNOWN';
  nhsAppTemplateName: string;
  nhsAppTemplateMessage: string;
};

export type FormState = {
  validationError?: FormErrorState;
  redirect?: string;
};

export type TemplateFormState = FormState & Session;

export type PageProps = {
  params: {
    sessionId: string;
  };
};

export type TemplateSubmittedPageProps = {
  params: {
    templateId: string;
  };
};

export type PageComponentProps = {
  initialState: TemplateFormState;
};

export type SubmitTemplatePageComponentProps = {
  templateName: string;
  sessionId: string;
};

export enum TemplateFormatText {
  APP = 'APP',
  EMAIL = 'EMAIL',
  LETTER = 'LETTER',
  SMS = 'SMS',
}
