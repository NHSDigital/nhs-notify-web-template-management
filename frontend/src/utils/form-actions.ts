'use server';

import { getSessionServer } from '@utils/amplify-utils';
import {
  CreateUpdateTemplate,
  isTemplateDtoValid,
  RoutingConfig,
  RoutingConfigStatusActive,
  routingConfigurationApiClient,
  TemplateDto,
  ValidatedTemplateDto,
} from 'nhs-notify-backend-client';
import { logger } from 'nhs-notify-web-template-management-utils/logger';
import { templateApiClient } from 'nhs-notify-backend-client/src/template-api-client';
import { sortAscByCreatedAt } from './sort';

export async function createTemplate(
  template: CreateUpdateTemplate
): Promise<TemplateDto> {
  const { accessToken } = await getSessionServer();

  if (!accessToken) {
    throw new Error('Failed to get access token');
  }

  const { data, error } = await templateApiClient.createTemplate(
    template,
    accessToken
  );

  if (error) {
    logger.error('Failed to create template', { error });
    throw new Error('Failed to create new template');
  }

  return data;
}

export async function uploadLetterTemplate(
  template: CreateUpdateTemplate,
  pdf: File,
  csv: File
) {
  const { accessToken } = await getSessionServer();

  if (!accessToken) {
    throw new Error('Failed to get access token');
  }

  const { data, error } = await templateApiClient.uploadLetterTemplate(
    template,
    accessToken,
    pdf,
    csv?.size > 0 ? csv : undefined
  );

  if (error) {
    logger.error('Failed to create letter template', { error });
    throw new Error('Failed to create new letter template');
  }

  return data;
}

export async function saveTemplate(
  template: Extract<TemplateDto, { templateType: 'EMAIL' | 'SMS' | 'NHS_APP' }>
): Promise<TemplateDto> {
  const { accessToken } = await getSessionServer();

  if (!accessToken) {
    throw new Error('Failed to get access token');
  }

  const { data, error } = await templateApiClient.updateTemplate(
    template.id,
    template,
    accessToken
  );

  if (error) {
    logger.error('Failed to save template', { error });
    throw new Error('Failed to save template data');
  }

  return data;
}

export async function setTemplateToSubmitted(
  templateId: string
): Promise<TemplateDto> {
  const { accessToken } = await getSessionServer();

  if (!accessToken) {
    throw new Error('Failed to get access token');
  }

  const { data, error } = await templateApiClient.submitTemplate(
    templateId,
    accessToken
  );

  if (error) {
    logger.error('Failed to save template', { error });
    throw new Error('Failed to save template data');
  }

  return data;
}

export async function setTemplateToDeleted(templateId: string): Promise<void> {
  const { accessToken } = await getSessionServer();

  if (!accessToken) {
    throw new Error('Failed to get access token');
  }

  const { error } = await templateApiClient.deleteTemplate(
    templateId,
    accessToken
  );

  if (error) {
    logger.error('Failed to save template', { error });
    throw new Error('Failed to save template data');
  }
}

export async function requestTemplateProof(
  templateId: string
): Promise<TemplateDto> {
  const { accessToken } = await getSessionServer();

  if (!accessToken) {
    throw new Error('Failed to get access token');
  }

  const { data, error } = await templateApiClient.requestProof(
    templateId,
    accessToken
  );

  if (error) {
    logger.error('Failed to request proof', { error });
    throw new Error('Failed to request proof');
  }

  return data;
}

export async function getTemplate(
  templateId: string
): Promise<TemplateDto | undefined> {
  const { accessToken } = await getSessionServer();

  if (!accessToken) {
    throw new Error('Failed to get access token');
  }

  const { data, error } = await templateApiClient.getTemplate(
    templateId,
    accessToken
  );

  if (error) {
    logger.error('Failed to get template', { error });
  }

  return data;
}

export async function getTemplates(): Promise<TemplateDto[]> {
  const { accessToken } = await getSessionServer();

  if (!accessToken) {
    throw new Error('Failed to get access token');
  }

  const { data, error } = await templateApiClient.listTemplates(accessToken);

  if (error) {
    logger.error('Failed to get templates', { error });
    return [];
  }

  const sortedData = data
    .map((template) => isTemplateDtoValid(template))
    .filter(
      (template): template is ValidatedTemplateDto => template !== undefined
    );

  return sortAscByCreatedAt(sortedData);
}

export async function getRoutingConfigs(): Promise<RoutingConfig[]> {
  const { accessToken } = await getSessionServer();

  if (!accessToken) {
    throw new Error('Failed to get access token');
  }

  const { data, error } = await routingConfigurationApiClient.list(accessToken);

  if (error) {
    logger.error('Failed to get routing configuration', {
      error: error,
    });
    return [];
  }

  return sortAscByCreatedAt(data);
}

export async function countRoutingConfigs(
  status: RoutingConfigStatusActive
): Promise<number> {
  const { accessToken } = await getSessionServer();

  if (!accessToken) {
    throw new Error('Failed to get access token');
  }

  const { data, error } = await routingConfigurationApiClient.count(
    accessToken,
    status
  );

  if (error) {
    logger.error(`Failed to count routing configuration for ${status}`, {
      error,
    });
    return 0;
  }

  return data.count;
}
