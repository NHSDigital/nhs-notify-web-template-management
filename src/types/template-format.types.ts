export enum TemplateFormatText {
  APP = 'APP',
  EMAIL = 'EMAIL',
  LETTER = 'LETTER',
  SMS = 'SMS',
}
