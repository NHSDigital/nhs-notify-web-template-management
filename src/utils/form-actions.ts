'use server';

import { getAmplifyBackendClient } from '@utils/amplify-utils';
import { DbOperationError } from '@domain/errors';
import { Template, Draft } from './types';
import { logger } from './logger';

export async function createTemplate(
  template: Draft<Template>
): Promise<Template> {
  const { data, errors } =
    await getAmplifyBackendClient().models.TemplateStorage.create(template);

  if (errors || !data) {
    logger.error('Failed to create template', errors);
    throw new Error('Failed to create new template');
  }
  return data;
}

export async function saveTemplate(template: Template): Promise<Template> {
  const { data, errors } =
    await getAmplifyBackendClient().models.TemplateStorage.update(template);

  if (errors) {
    logger.error('Failed to save template', errors);
    throw new Error('Failed to save template data');
  }

  if (!data) {
    throw new DbOperationError({
      message:
        'Template in unknown state. No errors reported but entity returned as falsy',
      operation: 'update',
    });
  }
  return data;
}

export async function getTemplate(
  templateId: string
): Promise<Template | undefined> {
  const { data, errors } =
    await getAmplifyBackendClient().models.TemplateStorage.get({
      id: templateId,
    });

  if (errors) {
    logger.error('Failed to get template', errors);
  }

  if (!data) {
    logger.warn(`Failed to retrieve template for ID ${templateId}`);
    return undefined;
  }

  return data;
}

export async function sendEmail(
  templateId: string,
  templateName: string,
  templateMessage: string,
  templateSubjectLine: string | null
) {
  const res = await getAmplifyBackendClient().queries.sendEmail({
    recipientEmail: 'england.test.cm@nhs.net',
    templateId,
    templateName,
    templateMessage,
    templateSubjectLine,
  });

  if (res.errors) {
    logger.error({
      description: 'Error sending email',
      res,
    });
  }
}

export async function getTemplates(): Promise<Template[] | []> {
  const { data, errors } =
    await getAmplifyBackendClient().models.TemplateStorage.list();

  if (errors) {
    logger.error('Failed to get templates', errors);
  }

  if (!data) {
    logger.warn(`Failed to retrieve templates`);
    return [];
  }

  return data;
}
