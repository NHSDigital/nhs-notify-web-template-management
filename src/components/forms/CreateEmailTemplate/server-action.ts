import { TemplateFormState } from '@utils/types';
import { zodValidationServerAction } from '@utils/zod-validation-server-action';
import { z } from 'zod';
import { saveSession } from '@utils/form-actions';

const schema = z.object({
  formId: z.enum(['create-email-template-back', 'create-email-template'], {
    message: 'Internal server error',
  }),
});

const formIdToServerActionMap: Record<
  'create-email-template-back' | 'create-email-template',
  (formState: TemplateFormState, formData: FormData) => TemplateFormState
> = {
  'create-email-template-back': (formState, formData) =>
    zodValidationServerAction(
      formState,
      formData,
      z.object({
        emailTemplateName: z.string({ message: 'Internal server error' }),
        emailTemplateSubjectLine: z.string({
          message: 'Internal server error',
        }),
        emailTemplateMessage: z.string({ message: 'Internal server error' }),
      })
    ),
  'create-email-template': (formState: TemplateFormState, formData: FormData) =>
    zodValidationServerAction(
      formState,
      formData,
      z.object({
        emailTemplateName: z
          .string()
          .min(1, { message: 'Enter a template name' }),
        emailTemplateSubjectLine: z
          .string()
          .min(1, { message: 'Enter a template subject line' }),
        emailTemplateMessage: z
          .string()
          .min(1, { message: 'Enter a template message' }),
      })
    ),
};

const formIdToPageMap: Record<
  'create-email-template-back' | 'create-email-template',
  string
> = {
  'create-email-template-back': 'choose-a-template-type',
  'create-email-template': 'preview-email-template',
};

export async function createEmailTemplateAction(
  formState: TemplateFormState,
  formData: FormData
): Promise<TemplateFormState> {
  const formId = formData.get('form-id');
  const parsedFormId = schema.safeParse({ formId });

  if (!parsedFormId.success) {
    return {
      ...formState,
      validationError: parsedFormId.error.flatten(),
    };
  }

  const response = formIdToServerActionMap[parsedFormId.data.formId](
    formState,
    formData
  );

  if (!response.validationError) {
    await saveSession(response);

    const page = formIdToPageMap[parsedFormId.data.formId];
    return {
      ...response,
      redirect: `/${page}/${formState.id}`,
    };
  }

  return response;
}
