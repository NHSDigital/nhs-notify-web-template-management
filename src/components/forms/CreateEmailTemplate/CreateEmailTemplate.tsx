'use client';

// we need this to be a client component because nhsuk-react-components uses client-only react features

import { FC } from 'react';
import { useFormState } from 'react-dom';
import {
  TextInput,
  HintText,
  Label,
  Textarea,
  Button,
} from 'nhsuk-react-components';
import { processFormActions } from '@forms/CreateEmailTemplate/server-action';
import { ZodErrorSummary } from '@molecules/ZodErrorSummary/ZodErrorSummary';
import { NHSNotifyFormWrapper } from '@molecules/NHSNotifyFormWrapper/NHSNotifyFormWrapper';
import { NHSNotifyBackButton } from '@molecules/NHSNotifyBackButton/NHSNotifyBackButton';
import { TemplateNameGuidance } from '@molecules/TemplateNameGuidance';
import { Personalisation } from '@molecules/Personalisation/Personalisation';
import { MessageFormatting } from '@molecules/MessageFormatting/MessageFormatting';
import { PageComponentProps, Template } from '@utils/types';
import { TemplateType } from '@utils/enum';
import { createEmailTemplatePageContent } from '@content/content';
import { FormSection } from '@molecules/FormSection/FormSection';
import { useTextInput } from '@hooks/use-text-input.hook';

export const CreateEmailTemplate: FC<PageComponentProps<Template>> = ({
  initialState,
}) => {
  const {
    pageHeading,
    errorHeading,
    buttonText,
    templateNameLabelText,
    templateSubjectLineLabelText,
    templateMessageLabelText,
    templateNameHintText,
  } = createEmailTemplatePageContent;

  const [state, action] = useFormState(processFormActions, initialState);

  const [emailTemplateName, emailTemplateNameHandler] =
    useTextInput<HTMLInputElement>(state.EMAIL?.name ?? '');

  const [emailTemplateSubjectLine, emailTemplateSubjectLineHandler] =
    useTextInput<HTMLInputElement>(state.EMAIL?.subject ?? '');

  const [emailTemplateMessage, emailTemplateMessageHandler] =
    useTextInput<HTMLTextAreaElement>(state.EMAIL?.message ?? '');

  const templateNameError =
    state.validationError?.fieldErrors.emailTemplateName?.join(', ');

  const templateSubjectLineError =
    state.validationError?.fieldErrors.emailTemplateSubjectLine?.join(', ');

  const templateMessageError =
    state.validationError?.fieldErrors.emailTemplateMessage?.join(', ');

  return (
    <div className='nhsuk-grid-row'>
      <NHSNotifyBackButton formId='create-email-template-back' action={action}>
        <input
          type='hidden'
          name='emailTemplateName'
          value={emailTemplateName}
        />
        <input
          type='hidden'
          name='emailTemplateSubjectLine'
          value={emailTemplateSubjectLine}
        />
        <input
          type='hidden'
          name='emailTemplateMessage'
          value={emailTemplateMessage}
        />
      </NHSNotifyBackButton>
      <div className='nhsuk-grid-column-two-thirds'>
        <ZodErrorSummary errorHeading={errorHeading} state={state} />
        <NHSNotifyFormWrapper action={action} formId='create-email-template'>
          <h1 className='nhsuk-heading-xl' data-testid='page-heading'>
            {pageHeading}
          </h1>

          <FormSection>
            <div className={templateNameError && 'nhsuk-form-group--error'}>
              <Label htmlFor='emailTemplateName'>{templateNameLabelText}</Label>
              <HintText>{templateNameHintText}</HintText>
              <TemplateNameGuidance template={TemplateType.EMAIL} />
              <TextInput
                id='emailTemplateName'
                onChange={emailTemplateNameHandler}
                value={emailTemplateName}
                error={templateNameError}
                errorProps={{ id: 'emailTemplateName-error-message' }}
                data-testid='emailTemplateName-input'
              />
            </div>
          </FormSection>

          <FormSection>
            <div
              className={templateSubjectLineError && 'nhsuk-form-group--error'}
            >
              <Label htmlFor='emailTemplateSubjectLine'>
                {templateSubjectLineLabelText}
              </Label>
              <TextInput
                id='emailTemplateSubjectLine'
                onChange={emailTemplateSubjectLineHandler}
                value={emailTemplateSubjectLine}
                error={templateSubjectLineError}
                errorProps={{ id: 'emailTemplateSubjectLine-error-message' }}
                data-testid='emailTemplateSubjectLine-input'
              />
            </div>

            <Textarea
              label={templateMessageLabelText}
              id='emailTemplateMessage'
              rows={10}
              onChange={emailTemplateMessageHandler}
              value={emailTemplateMessage}
              error={templateMessageError}
              errorProps={{ id: 'emailTemplateMessage-error-message' }}
              data-testid='emailTemplateMessage-input'
            />
          </FormSection>
          <Button type='submit' id='create-email-template-submit-button'>
            {buttonText}
          </Button>
        </NHSNotifyFormWrapper>
      </div>
      <div className='nhsuk-grid-column-one-third'>
        <Personalisation />
        <MessageFormatting template={TemplateType.EMAIL} />
      </div>
    </div>
  );
};
