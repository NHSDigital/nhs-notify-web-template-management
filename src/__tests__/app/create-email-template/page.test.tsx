import CreateEmailTemplatePage from '@app/create-email-template/[sessionId]/page';
import { getSession } from '@utils/form-actions';
import { Session, TemplateType } from '@utils/types';
import { redirect } from 'next/navigation';
import { CreateEmailTemplate } from '@forms/CreateEmailTemplate/CreateEmailTemplate';

jest.mock('@utils/form-actions');
jest.mock('next/navigation');
jest.mock('@forms/CreateEmailTemplate/CreateEmailTemplate');

const getSessionMock = jest.mocked(getSession);
const redirectMock = jest.mocked(redirect);
const CreateEmailTemplateMock = jest.mocked(CreateEmailTemplate);

const initialState: Session = {
  id: 'session-id',
  templateType: TemplateType.EMAIL,
  nhsAppTemplateName: '',
  nhsAppTemplateMessage: '',
};

describe('CreateEmailTemplatePage', () => {
  beforeEach(jest.resetAllMocks);

  it('should redirect to invalid-session when no session is found', async () => {
    getSessionMock.mockResolvedValueOnce(undefined);

    await CreateEmailTemplatePage({ params: { sessionId: 'session-id' } });

    expect(getSessionMock).toHaveBeenCalledWith('session-id');

    expect(redirectMock).toHaveBeenCalledWith('/invalid-session', 'replace');
  });

  it('should redirect to invalid-session when sessions template type is not EMAIL', async () => {
    getSessionMock.mockResolvedValueOnce({
      ...initialState,
      templateType: TemplateType.NHS_APP,
    });

    await CreateEmailTemplatePage({ params: { sessionId: 'session-id' } });

    expect(getSessionMock).toHaveBeenCalledWith('session-id');

    expect(redirectMock).toHaveBeenCalledWith('/invalid-session', 'replace');
  });

  it('should render CreateEmailTemplatePage component when session is found', async () => {
    getSessionMock.mockResolvedValueOnce(initialState);
    CreateEmailTemplateMock.mockImplementationOnce(() => <p>rendered</p>);

    const page = await CreateEmailTemplatePage({
      params: { sessionId: 'session-id' },
    });

    expect(getSessionMock).toHaveBeenCalledWith('session-id');

    expect(page).toMatchSnapshot();
  });
});
