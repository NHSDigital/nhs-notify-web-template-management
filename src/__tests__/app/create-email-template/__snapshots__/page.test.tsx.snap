// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateEmailTemplatePage 1`] = `
<DocumentFragment>
  <h2
    class="nhsuk-heading-l"
    data-testid="page-sub-heading"
  >
    Placeholder email page
  </h2>
</DocumentFragment>
`;
